packages/rpc-transport/src/transports/websocket/websocket-connection.ts

```diff
@@ -54,11 +60,48 @@ export async function createWebSocketConnection({ signal, url }: Config): Promis
             }
         }
         let hasConnected = false;
+        let bufferDrainWatcher: Readonly<{ onCancel(): void; promise: Promise<void> }> | undefined;
         function handleOpen() {
             hasConnected = true;
             resolve({
                 async send(payload: unknown) {
                     const message = JSON.stringify(payload);
+                    if (
+                        !bufferDrainWatcher &&
+                        webSocket.readyState === WebSocket.OPEN &&
+                        webSocket.bufferedAmount > sendBufferHighWatermark
+                    ) {
+                        let onCancel: () => void;
+                        const promise = new Promise<void>((resolve, reject) => {
+                            const intervalId = setInterval(() => {
+                                if (
+                                    webSocket.readyState !== WebSocket.OPEN ||
+                                    !(webSocket.bufferedAmount > sendBufferHighWatermark)
+                                ) {
+                                    clearInterval(intervalId);
+                                    bufferDrainWatcher = undefined;
+                                    resolve();
+                                }
+                            }, 16);
+                            onCancel = () => {
+                                bufferDrainWatcher = undefined;
+                                clearInterval(intervalId);
+                                reject(
+                                    // TODO: Coded error
+                                    new Error('WebSocket was closed before payload could be sent')
+                                );
+                            };
+                        });
+                        bufferDrainWatcher = {
+                            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                            // @ts-ignore
+                            onCancel,
+                            promise,
```

Comment:
TypeScript yells at you for this?

Reply:
The problem here is that:

1. Promise callbacks run synchonously
2. I know that
3. TypeScript pretends that it doesn't know that

If you search around their GitHub you'll find lots of complaints about that.